File: bumperbot_cpp_examples/src/simple_qos_subscriber.cpp
#include <rclcpp/rclcpp.hpp> //Include the ros C++ library
#include <std_msgs/msg/string.hpp> //Include the standard message type string

using std::placeholders::_1;

class SimpleQoSSubscriber : public rclcpp::Node
{
public:
    SimpleQoSSubscriber() : Node("simple_qos_subscriber"), qos_profile_sub_(10)
    {
        declare_parameter<std::string>("reliability", "system_default"); //initiation of parameter delcaration
        declare_parameter<std::string>("durability", "system_default");

        const auto reliability = get_parameter("reliability").as_string(); //variable setting of parameter
        const auto durability = get_parameter("durability").as_string();

        //setting up reliability as per defined
        if(reliability == "best_effort"){
            qos_profile_sub_.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", reliability.c_str());
        }
        else if(reliability == "reliable"){
            qos_profile_sub_.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", reliability.c_str());
        }
        else if(reliability == "system_default"){
            qos_profile_sub_.reliability(RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", reliability.c_str());
        }
        else {
            RCLCPP_ERROR_STREAM(get_logger(), "Selected Reliability QoS: " << reliability << " doesn't exists!");
        }

        //setting up durability as per defined
        if(durability == "volatile"){
            qos_profile_sub_.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", durability.c_str());
        }
        else if(durability == "transient_local"){
            qos_profile_sub_.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", durability.c_str());
        }
        else if(durability== "system_default"){
            qos_profile_sub_.durability(RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT);
            RCLCPP_INFO(get_logger(), "[Reliability] : %s", durability.c_str());
        }
        else {
            RCLCPP_ERROR_STREAM(get_logger(), "Selected Reliability QoS: " << durability << " doesn't exists!");
        }

        sub_ = create_subscription<std_msgs::msg::String>("chatter", qos_profile_sub_, std::bind(&SimpleQoSSubscriber::msgCallback, this, _1));
    }

private:
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
    rclcpp::QoS qos_profile_sub_;

    void msgCallback(const std_msgs::msg::String &msg) const
    {
        RCLCPP_INFO_STREAM(get_logger(), "I heard: " << msg.data.c_str());
    }
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleQoSSubscriber>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_python_examples/bumperbot_python_examples/simple_qos_subscriber.py
import rclpy #ros2 library
from rclpy.node import Node

from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy
from std_msgs.msg import String #Library for Message type which will be recieved

class SimpleQoSSubscriber(Node):
    def __init__(self):
        super().__init__("simple_qos_subscriber")

        self.qos_profile_sub = QoSProfile(depth=10) #Make a quality of service profile

        self.declare_parameter("reliability", "system_default") #Here reliability parameter is set to system default
        self.declare_parameter("durability", "system_default") #Here durability parameter is set to system default

        reliability = self.get_parameter("reliability").get_parameter_value().string_value
        durability = self.get_parameter("durability").get_parameter_value().string_value

        if reliability == "best_effort":
            self.qos_profile_sub.reliability = QoSReliabilityPolicy.BEST_EFFORT
            self.get_logger().info(f"[Reliability] : {reliability}")
        elif reliability == "reliable":
            self.qos_profile_sub.reliability = QoSReliabilityPolicy.RELIABLE
            self.get_logger().info(f"[Reliability] : {reliability}")
        elif reliability == "system_default":
            self.qos_profile_sub.reliability = QoSReliabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info(f"[Reliability] : {reliability}")
        else:
            self.get_logger().error(f"Selected Reliability QoS \"{reliability}\" doesn't exist")
            return
        
        if durability == "volatile":
            self.qos_profile_sub.durability = QoSDurabilityPolicy.VOLATILE
            self.get_logger().info(f"[Durability] : {durability}")
        elif durability == "transient_local":
            self.qos_profile_sub.durability = QoSDurabilityPolicy.TRANSIENT_LOCAL
            self.get_logger().info(f"[Durability] : {durability}")
        elif durability == "system_default":
            self.qos_profile_sub.durability = QoSDurabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info(f"[Durability] : {durability}")
        else:
            self.get_logger().error(f"Selected Durability QoS \"{durability}\" doesn't exist")
            return

        self.sub_ = self.create_subscription(String, "chatter", self.msgCallback, self.qos_profile_sub)

    def msgCallback(self, msg):
        self.get_logger().info(f"I heard: {msg.data}")

def main():
    rclpy.init()
    simple_qos_subscriber = SimpleQoSSubscriber()
    rclpy.spin(simple_qos_subscriber)
    SimpleQoSSubscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()